### PR TITLE
Add support for building merkle multiproofs

### PR DESCRIPTION
**Pull request by etan-status**
This adds functionality for building merkle multiproofs in the same form
that the Ethereum consensus specs suggest, i.e., in descending order of
helper indices. Merkle multiproofs are useful for the light client sync.
https://github.com/ethereum/consensus-specs/blob/v1.1.6/ssz/merkle-proofs.md#merkle-multiproofs

### CODE REVIEW

**Review by zah — COMMENTED** [review]
How come there are no tests here?

File: ssz_serialization/merkleization.nim
Comment (etan-status) [reply ssz_serialization/merkleization.nim:532]: This was moved from the original code:

```
if arr.len == 0:
   return getFinalHash(merkleizer)
```

Agree though that having the correct `zeroHash` directly is clearer. Will change it.

File: ssz_serialization/merkleization.nim
Comment (etan-status) [reply ssz_serialization/merkleization.nim:1111]: Good catch! I'm going with the first version, so we can benefit also from the code-size reduction from using the templates in this case. The templates are not exported.

File: ssz_serialization/merkleization.nim
Comment (etan-status) [reply ssz_serialization/merkleization.nim:477]: `o` is offset, and `n` is num items to include. Replaced with `firstIdx` and `numFromFirst`.

And replaced `Limit` with `int` which seems to be more appropriate as of `proc len*[TOpenArray: openArray|varargs](x: TOpenArray): int`

**Comment by etan-status** [comment]
Added tests.

File: ssz_serialization/merkleization.nim
Comment (etan-status) [reply ssz_serialization/merkleization.nim:477]: Actually, I think `Limit` is appropriate, to cover the case of arrays longer than 32-bit on 32-bit OS.

**Comment by etan-status** [comment]
New test coverage:
```
rm -rf coverage && mkdir -p coverage && nim c -d:PREFER_BLST_SHA256=false --nimcache:coverage/nimcache --passC:-fprofile-arcs --passC:-ftest-coverage --passL:-fprofile-arcs --passL:-ftest-coverage -r vendor/nim-ssz-serialization/tests/test_merkleization_types.nim && lcov --capture --directory coverage/nimcache --output-file coverage/coverage.info && lcov --extract coverage/coverage.info "*/vendor/nim-ssz-serialization/*" --output-file coverage/coverage.f.info && genhtml coverage/coverage.f.info --output-directory coverage/output
```

```
Overall coverage rate:
  lines......: 93.9% (939 of 1000 lines)
  functions..: 98.2% (161 of 164 functions)
```

File: ssz_serialization/merkleization.nim
Comment (etan-status) [inline ssz_serialization/merkleization.nim:518]: Handle case for arr.len > uint32.high on i386.

Comment (etan-status) [reply ssz_serialization/merkleization.nim:518]: NVM, that's not something that can happen ^^

**Comment by etan-status** [comment]
Wait for https://github.com/status-im/nim-ssz-serialization/pull/19 before merging this.

File: ssz_serialization/merkleization.nim
Comment (zah) [inline ssz_serialization/merkleization.nim:724]: It would be useful to provide some references to the spec being implemented here.

File: ssz_serialization/merkleization.nim
Comment (zah) [inline ssz_serialization/merkleization.nim:816]: Most of this logic is duplicated above and below. What are the differences?

File: ssz_serialization/merkleization.nim
Comment (zah) [inline ssz_serialization/merkleization.nim:1210]: Double evaluation of `v`

File: ssz_serialization/merkleization.nim
Comment (zah) [inline ssz_serialization/merkleization.nim:1335]: The code may be more clear if you place `type ResultType = Result[array[indices.len, Digest], string]` on the first line and then refer to it in the rest of the body.

File: ssz_serialization/proofs.nim
Comment (zah) [inline ssz_serialization/proofs.nim:236]: The `?` operator is not well defined in static context. It translates to a `return` statement propagating the error.

File: ssz_serialization/proofs.nim
Comment (zah) [inline ssz_serialization/proofs.nim:278]: I find the name `hash_tree_root` somewhat confusing for this use case. Why don't we use a more explicit name that references the generation of proofs.

Comment (etan-status) [reply ssz_serialization/proofs.nim:278]: A merkle proof is just a combination of roots, one for each `helper_index`. The `helper_indices` are computed with `get_helper_indices` from the requested `indices`, and then `hash_tree_root` is called on each of those `helper_indices` to compute the roots for each of them. To avoid O(n^2), all indices are passed to `hash_tree_root` in one go. However, `hash_tree_root` does not know how the roots will be used, I think the name is alright as it just gains an overload to compute multiple roots in one go, and being able to request non-1 generalized indices.

The function here is called `build_proof`, it's the same name that the EF consensus spec use for building proofs in their test suite.

We have `build_proof` overloads for dynamic data (e.g., client requested proofs), and also for static indices where helper indices can be pre-computed at compile time (e.g., light client update production). We also have `build_proof` overloads for cases where helper indices are already known, to avoid repeat computations.

Comment (zah) [reply ssz_serialization/proofs.nim:278]: I see, the explanation makes sense. Thanks!

File: ssz_serialization/merkleization.nim
Comment (etan-status) [reply ssz_serialization/merkleization.nim:724]: It is the same spec as the rest of the file is using, i.e., https://github.com/ethereum/consensus-specs/tree/dev/ssz

The only difference is that the overload allows querying for `GeneralizedIndex != 1`, as in getting intermediate hashes out.

File: ssz_serialization/proofs.nim
Comment (etan-status) [reply ssz_serialization/proofs.nim:236]: That is the intended behaviour here. The function would just statically become one that always tosses an error.

File: ssz_serialization/merkleization.nim
Comment (etan-status) [reply ssz_serialization/merkleization.nim:650]: Fixed by assigning `index` to a `let`. Also double checked the other templates; they do not share the problem.

File: ssz_serialization/merkleization.nim
Comment (etan-status) [reply ssz_serialization/merkleization.nim:724]: Spec link is actually already there, at the top of the file.

File: ssz_serialization/merkleization.nim
Comment (etan-status) [reply ssz_serialization/merkleization.nim:752]: Reason was consistency with the existing code above this hunk (~line 712):

```
      addField hashTreeRoot(toSszType(x.value))
```

I also don't understand why it's necessary there, figured it would be something weird like the dot-syntax not working in some non-testable scenarios, so did the same here.

It works fine for me without the `toSszType`, I have removed it here.

File: ssz_serialization/merkleization.nim
Comment (etan-status) [reply ssz_serialization/merkleization.nim:816]: Indeed, I have not found an elegant way around the duplication.

Four of the SSZ types get hashed according to the same format, with a subtree for the actual data to the left (`2.GeneralizedIndex`), and some metadata to the right (`3.GeneralizedIndex`).

- `BitList` (aka variable-sized `BitArray`) puts the members (padded with zeroes until capacity) in a subtree to the left (`2.GeneralizedIndex`), and put the current length to the right (`3.GeneralizedIndex`). It works using `bitListHashTreeRoot`.
- `List` (aka variable-sized `array`) puts the members (padded with zeroes until capacity) in a subtree to the left (`2.GeneralizedIndex`), and put the current length to the right (`3.GeneralizedIndex`). It works using `chunkedHashTreeRoot`. Compared to `BitList` it also takes into account that members are more complex than individual bits, so the code is not exactly same as in `BitList`.
- `HashList` (aka variable-sized `array` with a cache for intermediate hashes) puts the members (padded with zeroes until capacity) in a subtree to the left (`2.GeneralizedIndex`), and put the current length to the right (`3.GeneralizedIndex`). It works using `hashTreeRootCached`. Also calls directly into `chunkedHashTreeRoot` for small member sizes, so not exactly same as `BitList` / `List`.
- `SingleMemberUnion` puts the value in a subtree to the left (`2.GeneralizedIndex`), and puts the selector / union kind to the right (`3.GeneralizedIndex`). It works by hardcoding the selector to `0` as that is the only one supported for regular `hash_tree_root` by this codebase as well.

The code duplication mainly resembles the pre-existing code duplication for the existing `hash_tree_root` function. Sadly, they are all slightly different. Maybe some `template` shenanigans could avoid _some_ of the duplication, but it does not seem to make it easier for actually understanding it. As we have "only" four similar things in a file that rarely has to be checked, I went with the current approach of making the individual implementations look similar (so that once one of them is understood, all of them are understood), and just keeping the tiny differences in each of the four copies.

One way to review these segments is to search the file for `3.GeneralizedIndex` and then to keep swapping back and forth with `cmd-G` / `cmd-shift-G`. Editors like VSCode / CLion show the search results roughly at the same location of the screen when jumping around like this, allowing a visual comparison.

File: ssz_serialization/merkleization.nim
Comment (etan-status) [reply ssz_serialization/merkleization.nim:1210]: This template is only used by the `proc` immediately following it, and it is not passing computed expressions to the template. So, the double evaluation should be alright.

```
template normalize(v: GeneralizedIndex): GeneralizedIndex =
  ((v shl 1) or 1) shl leadingZeros(v)

proc cmpDepthFirst(x, y: GeneralizedIndex): int =
  cmp(x.normalize, y.normalize)
```

File: ssz_serialization/merkleization.nim
Comment (etan-status) [reply ssz_serialization/merkleization.nim:1335]: Oh, that's very clean indeed! Have switched to using this.

**Comment by etan-status** [comment]
`Nim devel` checks are failing because of upstream bugs in the Nim compiler repo.

**Comment by tersec** [comment]
> `Nim devel` checks are failing because of upstream bugs in the Nim compiler repo.

Is this an already-identified regression? One reason for having this Nim devel checks is so that they can be fixed upstream before Nimbus has to wait until Nim 1.x.2 or 1.x.3 before it even builds due to not having noticed these to be fixed in time for the Nim 1.x.0 or Nim 1.x.1 releases.

**Comment by stefantalpalaru** [comment]
> Is this an already-identified regression?

Yes, they fixed it here: https://github.com/nim-lang/Nim/pull/19472

**Comment by etan-status** [comment]
Re-ran CI.

**Comment by etan-status** [comment]
Anything still holding this PR back?

File: ssz_serialization/merkleization.nim
Comment (arnetheduck) [inline ssz_serialization/merkleization.nim:15]:
```suggestion
  std/[algorithm, sequtils],
```

Comment (etan-status) [reply ssz_serialization/merkleization.nim]: Done!

File: ssz_serialization/proofs.nim
Comment (kdeme) [inline ssz_serialization/proofs.nim:292]: Is there still a reason for having the function signatures with `proof: var openArray[Digest]` and a returned void Result (and the same counts for the `root_tree_hash` counterparts)? 

When there is already the need for `Result[void, string]` (or boolean for that matter), perhaps that type of return should be only provided, considering it is a safer practice (caller must verify the result to access the proofs).

**Comment by etan-status** [comment]
> I do have a similar remark as @zah regarding the naming of `hash_tree_root` procs which are there to return the proofs rather than the final hash tree root. I get it that each piece of the proof(s) is in its own a hash tree root, but based on the name, one could think that the final `hash_tree_root` call would then return the hash tree root, instead of the proof(s).

`hash_tree_root` does not compute a proof. It returns the root at index 1 by default, or the roots at custom given indices (like, if you pass 13, 42, the result will contain the root at index 13 and the root at index 42).
Note that the terminology of `root` for intermediate hashes, while it seems a bit backward, seems to be also used by the EF, e.g., see https://github.com/ethereum/consensus-specs/pull/2629#discussion_r716850110 -- I used to have this same concern as well.

Alternative names could be `hash_tree_root_at_indices`, `hash_tree_subroot` (but it can also do index 1), `hash_tree_leaves` (but it can also do intermediate hashes above the leaves). Or maybe there are better ideas?

> Initially I thought that the naming was fine, but then I noticed that the proof versions of these procs are actually all separate due to the indexing logic that needs to be done, which causes of no real code re-use with the original versions in there. Which is great from the point of view of leaving that code untouched and making merging this less impactful (Existing code touched is actually rather limited). But then I do think that for a lot of the code related to this indexing it would be more fitting to live in `proofs.nim`.

Yes, this is just an implementation detail though. The regular `hash_tree_root` is equivalent to `hash_tree_root(1.GeneralizedIndex)`, so it could also use the new implementation. However, due to the top root being requested much more frequently, I think having the 1 case implemented separately allows for more optimization. The cost is that for types where non-1 indices are requested that two copies of the hashing logic are emitted (one for top root, and one for any other index), but it's not many types that we use like that.

File: ssz_serialization/proofs.nim
Comment (etan-status) [reply ssz_serialization/proofs.nim:292]: I guess it could save some copying because the result can be directly put into the destination, but semantically these two are the same.
```
var foo: SomeObject
? state.build_proof(5.GeneralizedIndex, foo.proof)
? foo.proof = state.build_proof(5.GeneralizedIndex)
```

**Review by kdeme — APPROVED** [review]

**Pull request merged**
